internal/impl/ollama/completion_processor_ollama.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed as a Redpanda Enterprise file under the Redpanda Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+// https://github.com/redpanda-data/connect/blob/main/licenses/rcl.md
+
+package ollama
+
+import (
+	"context"
+	"errors"
+	"unicode/utf8"
+
+	"github.com/ollama/ollama/api"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+const (
+	ocpFieldPrompt = "prompt"
+)
+
+func init() {
+	err := service.RegisterBatchProcessor(
+		"ollama_completion",
+		ollamaCompletionProcessorConfig(),
+		ollamaCompletionBatchProcessor,
+	)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func ollamaCompletionProcessorConfig() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Fields(
+			service.NewStringField(bopFieldServerAddress).
+				Description("The address of a remote ollama server to use. If omitted a local ollama server will be started and ran.").
+				Optional(),
+			service.NewStringField(bopFieldModel).
+				Description("The ollama model to use").
+				Examples("llama3", "gemma2", "qwen2", "phi3"),
+			service.NewInterpolatedStringField(ocpFieldPrompt).
+				Description("The prompt to submit to the model. By default the entire payload as a string is submitted.").
+				Optional(),
+		)
+}
+
+func ollamaCompletionBatchProcessor(conf *service.ParsedConfig, mgr *service.Resources) (service.BatchProcessor, error) {
+	p := ollamaCompletionProcessor{}
+	if conf.Contains(ocpFieldPrompt) {
+		pf, err := conf.FieldInterpolatedString(ocpFieldPrompt)
+		if err != nil {
+			return nil, err
+		}
+		p.prompt = pf
+	}
+	b, err := newBaseProcessor(conf, mgr)
+	if err != nil {
+		return nil, err
+	}
+	p.baseOllamaProcessor = b
+	return &p, nil
+}
+
+type ollamaCompletionProcessor struct {
+	*baseOllamaProcessor
+
+	prompt *service.InterpolatedString
+}
+
+func (o *ollamaCompletionProcessor) ProcessBatch(ctx context.Context, batch service.MessageBatch) ([]service.MessageBatch, error) {
+	return o.baseOllamaProcessor.ProcessBatch(ctx, batch, func(ctx context.Context, msg *service.Message) (*service.Message, error) {
+		p, err := o.computePrompt(msg)
+		if err != nil {
+			return nil, err
+		}
+		g, err := o.generateCompletion(ctx, p)
+		if err != nil {
+			return nil, err
+		}
+		return service.NewMessage([]byte(g)), nil
```

Comment:
I'm not entirely sure what's the right thing to do here, but my impression is that if you just return a `service.NewMessage` here, you'll loose any metadata that's set on the origin message and I think most processors don't do that (should look that up, I'm not 100% sure).

Reply:
Thanks I looks at a couple other processors (redis, awk) and it seems they preserve metadata as well. I'm good with that - done!

On a side note - should the wasm processor be doing that too? Maybe not because the user is uploading code there and there are mappings for metadata?

Reply:
It's a tricky question, but if you look [here](https://github.com/redpanda-data/connect/blob/96f6b395e5e2cee71c0bced15ffeb1a575e015ab/internal/impl/wasm/functions.go#L74) and [here](https://github.com/redpanda-data/connect/blob/96f6b395e5e2cee71c0bced15ffeb1a575e015ab/internal/impl/wasm/functions.go#L97), you'll see utilities for reading and setting the metadata from the wasm processor. Users could run a wasm model which overwrites certain metadata fields in the current message, but they might want to preserve the rest. I can imagine that being relevant when processing Kafka messages for example and you may want at the output level to send them to different topics based on the `kafka_topic` metadata field. It would be annoying if the processor pipeline erased the metadata (although you can work around it in various ways, for example by using a `branch`).